![MEBS](./images/MEBS.png) 

# Welcome to the new location of MEBS 

The development of  https://github.com/eead-csic-compbio/metagenome_Pfam_score   and any updates and new versions of MEBS will be posted on this new website.


# Dependencies

- python >= 3.6.2
- python modules:
  - numpy
  - matplotlib
  - pandas
  - seaborn

# Basic usage

MEBS uses a few  command line options that can  be viewed by typing mebs.pl -h on the command line

```
perl mebs.pl -h 

  Program to compute MEBS for a set of genomic/metagenomic FASTA files in input folder.
  Version: v1.0

  usage: mebs.pl [options] 

   -help    Brief help message
   
   -input   Folder containing FASTA peptide files (.faa)                  (required)

   -type    Nature of input sequences, either 'genomic' or 'metagenomic'  (required)

   -fdr     Score cycles with False Discovery Rate 0.1 0.01 0.001 0.0001  (optional, default=0.01)

   -cycles  Show currently supported biogeochemical cycles
   

   
```

[Readme](https://eead-csic-compbio.github.io/metagenome_Pfam_score/READMEv1.html) needs to be updated 






























































